libnvme/ctrl.c

```diff
@@ -0,0 +1,439 @@
+_cnt_getter_func_gen_uint8_t(rab);
+_cnt_getter_func_gen_uint8_t(cmic);
+_cnt_getter_func_gen_uint8_t(mdts);
+_cnt_getter_func_gen_uint16_t(cntlid);
+_cnt_getter_func_gen_uint32_t(ver);
+_cnt_getter_func_gen_uint32_t(rtd3r);
+_cnt_getter_func_gen_uint32_t(rtd3e);
+_cnt_getter_func_gen_uint16_t(oaes);
+_cnt_getter_func_gen_uint16_t(oacs);
+_cnt_getter_func_gen_uint8_t(acl);
+_cnt_getter_func_gen_uint8_t(aerl);
+_cnt_getter_func_gen_uint8_t(frmw);
+_cnt_getter_func_gen_uint8_t(lpa);
+_cnt_getter_func_gen_uint8_t(elpe);
+_cnt_getter_func_gen_uint8_t(npss);
+_cnt_getter_func_gen_uint8_t(avscc);
+_cnt_getter_func_gen_uint8_t(apsta);
+_cnt_getter_func_gen_uint16_t(wctemp);
+_cnt_getter_func_gen_uint16_t(cctemp);
+_cnt_getter_func_gen_uint16_t(mtfa);
+_cnt_getter_func_gen_uint32_t(hmpre);
+_cnt_getter_func_gen_uint32_t(hmmin);
+_cnt_getter_func_gen_uint32_t(rpmbs);
+_cnt_getter_func_gen_uint16_t(edstt);
+_cnt_getter_func_gen_uint8_t(esto);
+_cnt_getter_func_gen_uint8_t(fwug);
+_cnt_getter_func_gen_uint16_t(kas);
+_cnt_getter_func_gen_uint16_t(hctma);
+_cnt_getter_func_gen_uint16_t(mntmt);
+_cnt_getter_func_gen_uint16_t(mxtmt);
+_cnt_getter_func_gen_uint32_t(sanicap);
+_cnt_getter_func_gen_uint8_t(sqes);
+_cnt_getter_func_gen_uint8_t(cqes);
+_cnt_getter_func_gen_uint16_t(maxcmd);
+_cnt_getter_func_gen_uint32_t(nn);
+_cnt_getter_func_gen_uint16_t(oncs);
+_cnt_getter_func_gen_uint16_t(fuses);
+_cnt_getter_func_gen_uint8_t(fna);
+_cnt_getter_func_gen_uint8_t(vwc);
+_cnt_getter_func_gen_uint16_t(awun);
+_cnt_getter_func_gen_uint16_t(awupf);
+_cnt_getter_func_gen_uint8_t(nvscc);
+_cnt_getter_func_gen_uint16_t(acwu);
+_cnt_getter_func_gen_uint32_t(sgls);
+_cnt_getter_func_gen_char(subnqn);
+_psd_getter_func_gen_uint16_t(mp);
+_psd_getter_func_gen_uint8_t(mxps);
+_psd_getter_func_gen_uint8_t(nops);
+_psd_getter_func_gen_uint32_t(enlat);
+_psd_getter_func_gen_uint32_t(exlat);
+_psd_getter_func_gen_uint8_t(rrt);
+_psd_getter_func_gen_uint8_t(rrl);
+_psd_getter_func_gen_uint8_t(rwt);
+_psd_getter_func_gen_uint8_t(rwl);
+_psd_getter_func_gen_uint16_t(idlp);
+_psd_getter_func_gen_uint8_t(ips);
+_psd_getter_func_gen_uint16_t(actp);
+_psd_getter_func_gen_uint8_t(apw);
+_psd_getter_func_gen_uint8_t(aps);
+
+static const char *_nvme_ctrl_data_prop_ascii(uint8_t *data, size_t size)
+{
+	char *rc_str = NULL;
+
+	assert(data != NULL);
+	assert(size != 0);
+
+	rc_str = (char *) calloc(sizeof(char), size + 1);
+	if (rc_str == NULL) {
+		errno = ENOMEM;
+		return NULL;
+	}
+
+	/* Remove trailing white space */
+	while(size > 0) {
+		if ((data[size - 1] == ' ') || (data[size - 1] == '\0'))
+			--size;
+		else
+			break;
+	}
+
+	if (size > 0)
+		memcpy(rc_str, data, size);
+
+	errno = 0;
+	return rc_str;
+}
+
+int _nvme_ctrl_new(const char *nvme_ctrl_dev_path, struct nvme_ctrl **cnt,
+		   char *err_msg)
+{
+	int rc = NVME_OK;
+	int fd = -1;
+
+	assert(nvme_ctrl_dev_path != NULL);
+	assert(cnt != NULL);
+
+	*cnt = (struct nvme_ctrl *) calloc(sizeof(struct nvme_ctrl), 1);
```

Comment:
The signature of calloc is `void *calloc(size_t nmemb, size_t size);`  The number of members/elements is first, the size of each element is second.